x-pack/plugin/watcher/src/internalClusterTest/java/org/elasticsearch/xpack/watcher/test/integration/SingleNodeTests.java

```diff
@@ -60,7 +63,12 @@ public void testThatLoadingWithNonExistingIndexWorks() throws Exception {
         assertThat(putWatchResponse.isCreated(), is(true));
 
         assertBusy(() -> {
-            indicesAdmin().prepareRefresh(".watcher-history*").get();
+            RefreshResponse refreshResponse = indicesAdmin().prepareRefresh(".watcher-history*").get();
+            assertThat(refreshResponse.getStatus(), equalTo(RestStatus.OK));
+            if (refreshResponse.getShardFailures().length > 0) {
+                logger.warn(refreshResponse.getShardFailures());
+            }
+            assertThat(refreshResponse.getSuccessfulShards(), equalTo(refreshResponse.getTotalShards()));
```

Comment:
Hm, looking at the failure, it'd seem that the actual search request failed with an exception:
```
SearchResponse searchResponse = client().prepareSearch(".watcher-history*").setSize(0).get();
``` 

```
at org.elasticsearch.xpack.watcher.test.integration.SingleNodeTests.lambda$testThatLoadingWithNonExistingIndexWorks$0(SingleNodeTests.java:64) |  
```

Also, note that the `assertBusy` only retries on `AssertionErrors`. It'll fail like all the other assertions on exceptions.

I don't think this will give us more information, the way it is now. 

I think we'd need to include the refresh/search calls into `assert...` methods (`assertAcked` ? ) so they throw `AssertionError` if we'd like to retry in case of an error.
Or maybe wait for the data stream to exist first? 
e.g. something along the lines of
```
  assertBusy(() -> assertTrue(dataStreamExists("watcher-history-16")));
```

Reply:
> Also, note that the assertBusy only retries on AssertionErrors. It'll fail like all the other assertions on exceptions.

Indeed, that's why I introduced assertion that the refresh succeeded on all shards. My assumption here is, that if a refresh succeeds then the shards are available so the search should succeed too. Do you think that assumption is wrong?

> Or maybe wait for the data stream to exist first?

I do like this idea because it's more explicit than checking the result of the refresh.

Reply:
>Indeed, that's why I introduced assertion that the refresh succeeded on all shards. My assumption here is, that if a refresh succeeds then the shards are available so the search should succeed too. Do you think that assumption is wrong?

I think this is what happened in the test run. 
Refresh succeeded and the (next) search call failed, if I'm reading the failure stack trace correctly.

Reply:
But if the shards are not available how can the refresh succeed and by succeed I mean in the response report that all shards have been successfully refreshed?

Reply:
refresh on missing targets reports `0` in the successful report
e.g. 
```
{
  "_shards": {
    "total": 0,
    "successful": 0,
    "failed": 0
  }
}
```

This wouldn't have caused an exception before this PR either.

When searching over a shard without a node in the routing ( https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/search/SearchShardIterator.java#L111 ) we do fail https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java#L327

IMO it's highly likely we would only need to wait for the `watch-history*` index to be `GREEN` before searching it

Reply:
Ah thank you for this clarification! I will change it right away.